### ISSUE TITLE
cce/8.7.6 compilation failures

### ISSUE THREAD

**Issue by jgphpc** [OpenACCUserGroup/OpenACCV-V#2]
Hi,

Trying to compile the [.c test files](https://github.com/OpenACCUserGroup/OpenACCV-V/tree/master/Tests) with cce/8.7.6, i am getting a few failures. It's probably fine because i expect cce to have only limited support for OpenACC>2.0 but i'd like to get your expert feedback about those failures, i.e can we explain those failure because cce just does not support the directives used in the test code (then i'll be fine with it). If not, we may want to report to the compiler team or adapt the src code.

## CC-10321

For instance:

> cc -hacc -hnoomp    atomic_capture_expr_bitand_x.c   -o atomic_capture_expr_bitand_x

```
CC-10321 craycc: ERROR File = atomic_capture_expr_bitand_x.c, Line = 60
  The specified statement does not have a valid form for an OpenACC atomic directive.
                   b[x] = totals[x%(n/10 + 1)] = a[x] & totals[x%(n/10 + 1)];
```

- atomic_capture_bitand_equals
- atomic_capture_bitor_equals
- atomic_capture_bitxor_equals
- atomic_capture_divided_equals
- atomic_capture_expr_bitand_x
- atomic_capture_expr_bitor_x
- atomic_capture_expr_bitxor_x
- atomic_capture_expr_divided_x
- atomic_capture_expr_lshift_x
- atomic_capture_expr_minus_x
- atomic_capture_expr_multiply_x
- atomic_capture_expr_plus_x
- atomic_capture_expr_rshift_x
- atomic_capture_lshift_equals
- atomic_capture_minus_equals
- atomic_capture_multiply_equals
- atomic_capture_plus_equals
- atomic_capture_postdecrement
- atomic_capture_postincrement
- atomic_capture_predecrement
- atomic_capture_preincrement
- atomic_capture_rshift_equals

## CC-10299

> cc -hacc -hnoomp    kernels_default_present.c   -o kernels_default_present

```
CC-10299 craycc: ERROR File = kernels_default_present.c, Line = 15
  The OpenACC default clause has a malformed argument.
      #pragma acc kernels default(present)
```

- kernels_default_present
- parallel_default_present

## CC-10302

> cc -hacc -hnoomp    parallel_loop_auto.c   -o parallel_loop_auto

```
CC-10302 craycc: ERROR File = parallel_loop_auto.c, Line = 18
  This OpenACC directive does not accept the "auto" clause.
        #pragma acc parallel loop auto
```

## undefined ref

> cc -hacc -hnoomp    acc_copyin_async.c   -o acc_copyin_async

```
CC-223 craycc: WARNING File = acc_copyin_async.c, Line = 19
  The function "acc_copyin_async" is declared implicitly.
      acc_copyin_async(a, n * sizeof(real_t), 1);
      ^


            for (int x = 0; x < n; ++x){
CC-7254 craycc: WARNING File = acc_copyin_async.c, Line = 26
   A loop which was marked for worksharing will not be workshared.

            for (int x = 0; x < n; ++x){
CC-7254 craycc: WARNING File = acc_copyin_async.c, Line = 33
   A loop which was marked for worksharing will not be workshared.

Cray C : Version 8.7.6 (20181015183930_cdba25e9872fe61c78223bf16dd108a77ffce687)
Total warnings detected in acc_copyin_async.c: 3
/tmp/pe_9764/acc_copyin_async_1.o: In function `test':
OpenACCV-V.git/Tests/acc_copyin_async.c:19: 
  undefined reference to `acc_copyin_async'
```

- acc_copyin_async
- acc_copyout_async
- acc_create_async
- acc_delete_async
- acc_get_default_async
- acc_memcpy_from_device_async
- acc_memcpy_to_device_async
- acc_copyin
- acc_create
- acc_malloc
- acc_set_default_async
- acc_update_device_async
- acc_update_self_async

**Comment by jgphpc** [comment]
For cce, i had to modify `acc_testsuite.h`:

```
diff --git a/Tests/acc_testsuite.h b/Tests/acc_testsuite.h
index f7a4861..3362ef0 100644
--- a/Tests/acc_testsuite.h
+++ b/Tests/acc_testsuite.h
@@ -26,7 +26,11 @@
 long long n = ARRAYSIZE;
 #define PRECISION   1e-8
 typedef double real_t;
+
+#ifndef _CRAYC
 typedef enum { false, true } bool;
+#endif
+
 #define ARRAYSIZE_NEW 1024
 #define ARRAYSIZE_SMALL 10
```

**Comment by utimatu** [comment]
CC-10321:
	I haven't taken the time to review all of these, but I cannot find any issues with the formats of the atomics.  The format for the capture in the specific one you have shown the error for is found on page 36 of the OpenACC specifications and is the last item in the list of valid capture statements: v = x = expr binop x. The atomic formats allowed in the spec are the same from 2.0 on, so it is probably an issue with Cray if they are not working.  If any of the tests specifically seem to have non-conforming forms of the atomic statement, please bring them to my attention.  It is tricky work to evaluate 300+ different valid formats of the atomic construct.  

CC-10299:
	The default(present) clause that it is complaining about was not included in the specifications until version 2.5, after Cray stopped updating for new OpenACC features, so the fact that it is not compiling happy is not surprising.  
	
CC-10302:
	This one I cannot explain.  The auto clause is valid on a loop construct, and since I don't see any restrictions otherwise, I assume that it can also be applied to a parallel loop construct.  It was in the specifications in 2.0 and was valid in that context.  My thoughts are that this might be a Cray issue. 
	
acc_copyin_async, acc_copyout_async, acc_create_async, acc_delete_async, acc_memcpy_from_device_async, acc_memcpy_to_device_async, acc_update_device_async, acc_update_self_async:
	These runtime routines were all added in the 2.5 version which added async data routines.  So I'm not surprised that Cray doesn't support these either.
	
acc_copyin, acc_create, acc_malloc:
	Each of these was already included in the OpenACC 2.0 spec.  I find it interesting that the three allocation routines are having problems but not the acc_copyout, acc_delete, or the acc_free.  Do you think you could give me the exact errors for these?  If the issue is undefined ref, what reference is undefined in these?  
	
acc_get_default_async, acc_set_default_async: 
	OpenACC 2.0 did not even have the acc-default-async-var ICV.  So having routines that could modify them would also not work properly with a pure 2.0 implementation.  
	

I really appreciate the work you did, and if you are willing, I would love to see the full list of results that you were able to gather on the Cray compiler.  As they do not support 2.5 and this project is a 2.5+ testsuite, we have not looked into the performance of Cray on the testsuite, but I would be curious to see the results anyway.

**Comment by ajarmusch** [comment]
Hi, 
Thank you for bringing this issue up at the time. Since Cray no longer supports C/C++ at the moment. We are going to close this issue and hope you will continue testing with the HPE Cray Fortran compiler as we are currently doing.
